llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -2065,14 +2065,26 @@ MachineBasicBlock *SIInstrInfo::insertSimulatedTrap(MachineRegisterInfo &MRI,
       .addImm(AMDGPU::SendMsg::ID_INTERRUPT);
   BuildMI(MBB, MI, DL, get(AMDGPU::S_MOV_B32), AMDGPU::M0)
       .addUse(AMDGPU::TTMP2);
-  BuildMI(MBB, MI, DL, get(AMDGPU::S_BRANCH)).addMBB(HaltLoop);
+
+  if (MBB.succ_empty()) {
+    BuildMI(MBB, MI, DL, get(AMDGPU::S_BRANCH)).addMBB(HaltLoop);
+  } else {
```

Comment:
kill and ENDPGM_TRAP have the same problem, but this handling looks different. Can you follow what they do?

Reply:
Sure, fixed in the update. Did you also mean using `s_cbranch_execnz`? I'm not certain why that's actually correct, can we just assume that exec is non-zero?

Reply:
Yes, that's part of it. If exec is zero no lanes are executing so no trap should happen

Reply:
Ah, so we should be guarding the whole trap sequence with a `c_branch_execnz`? I did this in the update. I think what was confusing me was that the other lowering for the trap intrinsic is an unconditional `s_trap 2`. I guess the argument is that should be okay since we don't delete the guarding `s_cbranch_execz` if it jumps over a trap [here](https://github.com/llvm/llvm-project/blob/8ef2011b2cd3a8fc2ef8d6ea0facb1a39a0dd621/llvm/lib/Target/AMDGPU/SIPreEmitPeephole.cpp#L327)?

Reply:
Now that I think about it, this should be unnecessary? By default we assume at least one line is active and have to insert branches over special instructions that should do nothing with exec 0, like trap

Reply:
I think we still need a "fake" branch so we don't disconnect definitions before the trap with their uses after it in the CFG (this was what I was trying to do with the `s_cmp %0,%0; s_cbranch_scc1` thing). I guess we should prefer c_branch_execnz since it seems like a more compact way of representing that. 

Incidentally, I guess [this comment ](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/SILowerControlFlow.cpp#L34)is not correct (the s_cbranch_execnz isn't "optional")?

Reply:
That comment is correct. In that context there's no special case instructions being skipped, like trap, that must be skipped if no lanes are active